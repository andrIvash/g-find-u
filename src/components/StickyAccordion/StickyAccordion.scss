@import '../../styles/abstracts/functions';
@import '../../styles/abstracts/breakpoints';

$wrapper_height: rem(400);
$padding: rem(12);

.s-acco {
    display: flex;
    flex-direction: column;
    height: $wrapper_height;
    background-color: lightgray;

    &--no-expanded {
        background-color: transparent;
    }
    .s-acco__item {
        border-top-width: 0;
        border-left-width: 0;
        border-bottom: 1px solid var(--app-border-color);
        background-color: var(--app-bgcolor);

        &:first-child, &:first-child.s-acco__item--expanded {
            border-top: 1px solid var(--app-border-color);
        }

        &.s-acco__item--expanded {
            flex: 1;
            margin: 0;
            border: 0;
            border-bottom-width: 0;

            .s-acco__sum {
                div {
                    margin: $padding 0;
                }
            }

            .s-acco__details {
                background-color: var(--app-ctcolor);
            }
        }

        .s-acco__details {
            background-color: var(--app-ctcolor);
        }

        .s-acco__sum {
            background-color: var(--app-bgcolor);
            border-bottom-color: var(--app-border-color);
            min-height: 0;
            margin: 0;

            & > div {
                border-bottom-width: 0;
            }
        }

        .s-acco__header {
            font-size: var(--app-font-block-size);
            font-family: var(--app-font-family);
            font-weight: normal;
            color: var(--app-color);
        }

        .s-acco__details {
            padding: 0 0 0 rem(16);
        }
    }
}

@media screen {
    @include break-min-max(rem(425), rem(1100)) {
        .s-acco {
            .s-acco__item {
                .s-acco__header {
                    font-size: var(--app-font-block-size-medium);
                }
            }
        }
    }

    @include break-min-max(rem(320), rem(500)) {
        .s-acco {
            .s-acco__item {
                .s-acco__header {
                    font-size: var(--app-font-block-size-small);
                }
            }
        }
    }
}