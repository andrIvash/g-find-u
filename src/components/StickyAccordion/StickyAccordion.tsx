import React, { useState, useMemo } from 'react';
import Accordion from '@mui/material/Accordion';
import AccordionDetails from '@mui/material/AccordionDetails';
import AccordionSummary from '@mui/material/AccordionSummary';
import Typography from '@mui/material/Typography';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import { IUser } from '../../store/users/types';
import { IProps } from './index';

import './StickyAccordion.scss';

export const StickyAccordion = (props: IProps) => {
    const {
        items,
        detailsHeight,
        renderContent,
        onChangeExpanded,
    } = props;
    const [expanded, setExpanded] = useState<string | false>();

    const handleChange = (login: string) => (_event: React.ChangeEvent<NonNullable<unknown>>, isExpanded: boolean) => {
        setExpanded(isExpanded ? login : false);
        if (onChangeExpanded) {
            onChangeExpanded(login);
        }
    };

    const resultContent = useMemo(() => {
        let result: IUser[] = [];
        const expandedItemId = items.findIndex((item) => item.login === expanded);
        if (expandedItemId === -1) {
            result = [...items];
        } else {
            if (items[expandedItemId - 1]) {
                result.push(items[expandedItemId - 1]);
            }
            result.push(items[expandedItemId]);
            if (items[expandedItemId + 1]) {
                result.push(items[expandedItemId + 1]);
            }
        }
        return result;
    }, [items, expanded]);

    return (
        <div className={`s-acco ${!expanded ? 's-acco--no-expanded' : ''}`}>
            {resultContent.map((item) => (
                <Accordion
                    className={`s-acco__item ${expanded === item.login ? 's-acco__item--expanded' : ''}`}
                    square
                    key={item.id}
                    expanded={expanded === item.login}
                    onChange={handleChange(item.login)}
                    variant='outlined'
                >
                    <AccordionSummary
                        className='s-acco__sum'
                        expandIcon={<ExpandMoreIcon />}
                        aria-controls={`${item.login}-content`}
                        id={`${item.login}-header`}
                    >
                        <Typography className='s-acco__header'>
                            {item.login}
                        </Typography>
                    </AccordionSummary>
                    <AccordionDetails
                        className='s-acco__details'
                        id={`scrollableDiv-${item.login}`}
                        style={{
                            height: detailsHeight ? detailsHeight : '100%',
                            overflow: detailsHeight ? 'auto' : undefined,
                        }}
                    >
                        {renderContent ?
                            renderContent(`scrollableDiv-${item.login}`, item.login) : undefined}
                    </AccordionDetails>
                </Accordion>
            ))}
        </div>
    );
};