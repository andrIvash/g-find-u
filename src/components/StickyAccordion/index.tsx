import { StickyAccordion } from "./StickyAccordion";
import { IUser } from '../../store/users/types';

export interface IProps {
    items: IUser[];
    detailsHeight?: number;
    onChangeExpanded?: (targetName: string) => void;
    renderContent?: (targetId: string, targetName: string) => JSX.Element | undefined;
}

export default StickyAccordion;