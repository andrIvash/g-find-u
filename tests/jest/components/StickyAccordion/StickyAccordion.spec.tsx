import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import StickyAccordion from '../../../../src/components/StickyAccordion';
import { usersMockData } from '../../../mocks/usersMockData';

describe('StickyAccordion Component', () => {

    it('renders without crashing', () => {
        render(<StickyAccordion items={usersMockData} />);
        expect(screen.getByText('User1')).toBeInTheDocument();
    });

    it('expands the correct accordion item on click', () => {
        render(<StickyAccordion items={usersMockData} />);
        fireEvent.click(screen.getByText('User1'));
        expect(screen.getByText('User1').closest('.Mui-expanded')).toBeInTheDocument();
    });

    it('calls onChangeExpanded prop when an accordion is expanded', () => {
        const mockOnChangeExpanded = jest.fn();
        render(<StickyAccordion items={usersMockData} onChangeExpanded={mockOnChangeExpanded} />);
        fireEvent.click(screen.getByText('User1'));
        expect(mockOnChangeExpanded).toHaveBeenCalledWith('User1');
    });

    it('renders custom content when renderContent is provided', () => {
        const mockRenderContent = jest.fn().mockReturnValue(<div>Custom Content</div>);
        render(<StickyAccordion items={usersMockData} renderContent={mockRenderContent} />);
        fireEvent.click(screen.getByText('User1'));
        expect(screen.getAllByText('Custom Content')).toHaveLength;
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<StickyAccordion items={usersMockData} />);
        expect(asFragment()).toMatchSnapshot();
    });
});