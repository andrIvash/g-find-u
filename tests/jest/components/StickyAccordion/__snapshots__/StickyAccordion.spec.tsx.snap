// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StickyAccordion Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="s-acco s-acco--no-expanded"
  >
    <div
      class="MuiPaper-root MuiPaper-outlined MuiAccordion-root MuiAccordion-gutters s-acco__item  css-1b9v8ew-MuiPaper-root-MuiAccordion-root"
      data-testid="accordion-item"
    >
      <div
        aria-controls="User1-content"
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters s-acco__sum css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
        id="User1-header"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 s-acco__header css-ahj2mt-MuiTypography-root"
            data-testid="accordion-itemheader"
          >
            User1
          </p>
        </div>
        <div
          class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              aria-labelledby="User1-header"
              class="MuiAccordion-region"
              id="User1-content"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root s-acco__details css-15v22id-MuiAccordionDetails-root"
                id="scrollableDiv-User1"
                style="height: 100%;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-outlined MuiAccordion-root MuiAccordion-gutters s-acco__item  css-1b9v8ew-MuiPaper-root-MuiAccordion-root"
      data-testid="accordion-item"
    >
      <div
        aria-controls="User2-content"
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters s-acco__sum css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
        id="User2-header"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 s-acco__header css-ahj2mt-MuiTypography-root"
            data-testid="accordion-itemheader"
          >
            User2
          </p>
        </div>
        <div
          class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              aria-labelledby="User2-header"
              class="MuiAccordion-region"
              id="User2-content"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root s-acco__details css-15v22id-MuiAccordionDetails-root"
                id="scrollableDiv-User2"
                style="height: 100%;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
